backend/src/main/java/com/google/minicrm/data/Lead.java

```diff
@@ -209,6 +245,74 @@ public static Key generateKey(Key parentKey, String leadId) {
     return KeyFactory.createKey(parentKey, KIND_NAME, leadId);
   }
 
+  /**
+   * Searches for any location data present and populates the estimatedLongitude and
+   * estimateLatitude instance variables.
+   */
+  private void generateLocationInfo() {
+    StringBuilder locationInfo = new StringBuilder();
+    String phoneNumber = "";
+
+    for (String key : columnData.keySet()) {
+      // If the column data is related with address add it into locationInfo arraylist
+      if (key.equals("POSTAL_CODE") || key.equals("STREET_ADDRESS") ||
+          key.equals("CITY") || key.equals("REGION") ||
+          key.equals("COUNTRY") || key.equals("COUNTRY")) {
+        locationInfo.append(columnData.get(key) + " ");
+      } else if (key.equals("PHONE_NUMBER") ) {
+        phoneNumber = columnData.get("PHONE_NUMBER");
+      }
+    }   
+    if (locationInfo.length() == 0) {
+     final int areaCodeFromLead = Integer.parseInt(phoneNumber.substring(0, 3));
+     System.out.println(areaCodeFromLead);
+
+     try {
+        // create Gson instance
+        Gson gson = new Gson();
+        // create a reader
+        ClassLoader classloader = Thread.currentThread().getContextClassLoader();
+        InputStream is = classloader.getResourceAsStream("data/AreaCodes.json");
+        InputStreamReader reader = 
+        new InputStreamReader(is, StandardCharsets.UTF_8);
+        
+        // convert JSON array to list of users
+        List<AreaCode> areaCodes = new Gson().fromJson(reader, new TypeToken<List<AreaCode>>() {}.getType());
+
+        List<AreaCode> areaCodesFiltered = areaCodes
+         .stream()
+         .filter(c -> c.areaCode == areaCodeFromLead)
+         .collect(Collectors.toList());
+        
+         for( AreaCode finalFilter : areaCodesFiltered) {
+         estimatedLatitude = finalFilter.latitude;
+         estimatedLongitude = finalFilter.longitude;
+         }
+            //close reader
+            reader.close();
+        } catch (Exception ex) {
```

Comment:
Is there a better exception to catch? (i.e. something more specific like IOException), also keep the try block smaller so that only the method(s) throwing the exception are included.

Reply:
+1